File: cpp/cpp_100-999/275.cc
// C++
// leetcode 275
// https://leetcode-cn.com/problems/h-index-ii/
// 经典二分

class Solution {
public:
    int hIndex(vector<int>& citations) {
    	int n = citations.size();
    	int h = INT_MIN;
    	int bad = INT_MAX;
    	int l = 1;
    	int r = n;
    	while(l <= r) {
    		int mid = l + (r - l);
    		if(citations[n - mid] < mid) {
                bad = mid;
    			r = mid -1;
    		} else {
    			h = mid;
    			l = mid + 1;
    		}
    	}
    	if(h == INT_MIN)
    		return bad - 1;
    	return h;
    }
};

File: Go/Go_100-999/275.go
// Go
// leetcode 275
// https://leetcode.cn/problems/h-index-ii/

func hIndex(citations []int) int {
	n := len(citations)
	l, r := 0, n-1
	for l<=r {
		mid := (l+r)/ 2
		if citations[mid] >= n-mid { // ok
			r = mid-1
		} else {
			l = mid+1
		}
	}
	return n-l
}